### DISCUSSION TITLE
Flexibility in defining policies

### DISCUSSION THREAD

**Discussion by elsalant** [fybrik/fybrik#579] [Developers]
@roee88 
I had been prototyping using the OPA-based katalog from a branch that Roee gave me.  This was great - I was able to extend the Asset template for my purposes, define whatever OPA policies I wanted, and, using a curl call, go directly to OPA from the module.  
For example, I need to do more than redaction, I want also to get a list of blocked URL based on a passed role and the policies.  
The M4D web page describing use of Rego (or at least the only one that I am aware of) is really short on explanations (https://607573df9860bf9afcf4805b--mesh-for-data.netlify.app/samples/notebook/)
Here is  part of the Rego that I previously had:
apiVersion: v1
data:
  main: |
    package katalog.example

    import data.kubernetes.assets

    filters[output] {
        count(rule)==0
        output = {"name": "Allow by default", "action": "Allow"}
    }

    filters[output] {
        count(rule)>0
        output = rule[_]
    }

    blockList[output] {
       count(blocked) == 0
       output = {"name":"Allow all URLs!", "action": "None"}
    }

    blockList[output] {
        count(blocked)>0
        output = blocked[_]
    }

   rule[{"name": "Full priviledges for Manager", "action": "Allow"}] {
        input.request.operation == "READ"
        input.request.role[_] = "Manager"
    }

    rule[{"name": "Redact PII columns if not Manager or HR for personnel data", "action": "RedactColumn", "columns": columns}] {
        not (input.request.role[_] = "Manager")
        not (input.request.role[_] = "HR")
        asset := assets[input.request.asset.namespace][input.request.asset.name]
        asset.spec.assetMetadata.tags[_] = "personnel"
        columns := [c | asset.spec.assetMetadata.componentsMetadata[i].tags[_] = "PII"; c = i]
    }

    blocked[{"name": "Block data-related tagged URLs for HR", "action": "BlockURL"}] {
       input.request.role[_] = "HR"
       asset := assets[input.request.asset.namespace][input.request.asset.name]
       asset.spec.assetMetadata.tags[_] = "data"
    }

    blocked[{"name": "Block order-related tagged URLs for all but Managers", "action": "BlockURL"}] {
       not (input.request.role[_] = "Manager")
       asset := assets[input.request.asset.namespace][input.request.asset.name]
       asset.spec.assetMetadata.tags[_] = "order"
    }

Note that there are two entry points, "blockList" - for determining if a requested URL should be blocked, and "filters" which returns a list of columns that need to be redacted based on a passed role.  

How can this be done now?  I really would like an interface to the Policy Manager where I pass the Rego query (e.g. "blockList" or "filters" as well as my own input parameters (such as role and organization which the module gets in run-time).
Additionally, in the prototype, the Rego was written into a configmap and it was possible to directly edit the yaml file.  Now, the only way I can see of doing this would be to dump the running configmap to a yaml file, edit that, and then reapply, but not only is this additional steps, but upon each change, I need to regenerate the yaml from the running configmap.  (E.g. kubectl get cm policy -n m4d-system -o yaml > policy.yaml).  Is there a better way of doing this?

Note that I try to use my own policy and having the module go directly to the OPA server, I get a strange error from the OPA server:
{"client_addr":"[::1]:48130","level":"info","msg":"Received request.","req_body":"package data_policies\n\ncorrect_input {\n\tcount(incorrect_input) == 0\n}\n\nincorrect_input[used_policy] {\n   not verify_access_type\n   used_policy := build_action_from_policies(build_policy_from_description(\"unknown access type\"))\n} {\n    not verify_purpose\n    used_policy := build_action_from_policies(build_policy_from_description(\"unknown purpose\"))\n} {\n    not verify_role\n    used_policy := build_action_from_policies(build_policy_from_description(\"unknown role\"))\n} {\n\tcheck_access_type([\"COPY\"])\n    not verify_geography\n    used_policy := build_action_from_policies(build_policy_from_description(\"unknown geography to copy the data\"))\n}","req_id":2844938,"req_method":"PUT","req_params":{},"req_path":"/v1/policies/m4d-system/opa-m4d-policy-lib/verify_correct_input.rego","time":"2021-05-05T09:22:38Z"}

**Comment by roee88** [comment]
It's hard to track what you're asking exactly. Check if this helps:
1. @ritwikc is working on how we use OPA and hopefully the result will be similar to what you want but it probably won't be the same
2. `kubectl edit` to edit the configmaps in the cluster. You can also put the rego directly in the configmap to begin with, no one forces you to create configmaps in just one particular way :)

Just to note for visibility that normally in M4D you don't interface with the policy manager directly, so your usage is special in the sense that it leverages the work we do with OPA directly.

**Reply by elsalant** [reply]
@ritwikc @RoeeShlomo 
Let me try to make what I need more succinct:
1. The ability to define the Rego OPA policies in any way that I want, including passing whatever parameters I want, and evaluating any of a number of defined Rego functions.  (I.e. first evaluate a function called, "blockList", then evaluate a function called "filterList" etc)
2. The ability to either pass to the Policy Manager any input values I want, and the name the OPA policy that I want to evaluate. (e.g "blockList", "filterList")
3. The ability to extend the Asset resource definition to match the requirements of my policies.

I have no idea what is going on now when I tried to define my own policies rules, and saw in the OPA log, "unknown access type, unknown role, unknown purpose"...
